plugins/plugin-docker/che-plugin-docker-compose/src/main/java/org/eclipse/che/plugin/docker/compose/yaml/ComposeEnvironmentParser.java

```diff
@@ -0,0 +1,173 @@
+     * @throws ServerException
+     *         when environment recipe can not be retrieved
+     */
+    public ComposeEnvironment parse(String recipeContent, String contentType) throws IllegalArgumentException,
+                                                                                         ServerException {
+        ComposeEnvironment composeEnvironment;
+        switch (contentType) {
+            case "application/x-yaml":
+            case "text/yaml":
+            case "text/x-yaml":
+                try {
+                    composeEnvironment = YAML_PARSER.readValue(recipeContent, ComposeEnvironment.class);
+                } catch (IOException e) {
+                    throw new IllegalArgumentException(
+                            "Parsing of environment configuration failed. " + e.getLocalizedMessage());
+                }
+                break;
+            default:
+                throw new IllegalArgumentException("Provided environment recipe content type '" +
+                                                   contentType +
+                                                   "' is unsupported. Supported values are: " +
+                                                   "application/x-yaml, text/yaml, text/x-yaml");
+        }
+        return composeEnvironment;
+    }
+    /**
+     * Converts Docker Compose environment model into YAML file.
+     *
+     * @param composeEnvironment Docker Compose environment model file
+     * @throws IllegalArgumentException
+     *         when argument is null or conversion to YAML fails
+     */
+    public String toYaml(ComposeEnvironment composeEnvironment) throws IllegalArgumentException {
+        checkNotNull(composeEnvironment, "Compose environment should not be null");
+        try {
+            return YAML_PARSER.writeValueAsString(composeEnvironment);
+        } catch (JsonProcessingException e) {
+            throw new IllegalArgumentException(e.getLocalizedMessage(), e);
+        }
+    }
+
+    private String getContentOfRecipe(EnvironmentRecipe environmentRecipe) throws ServerException {
+        if (environmentRecipe.getContent() != null) {
+            return environmentRecipe.getContent();
+        } else {
+            return recipeDownloader.getRecipe(environmentRecipe.getLocation());
+        }
+    }
+
+    private CheServicesEnvironmentImpl asCheEnvironment(ComposeEnvironment composeEnvironment) {
+        Map<String, CheServiceImpl> services = Maps.newHashMapWithExpectedSize(composeEnvironment.getServices().size());
+        for (Map.Entry<String, ComposeServiceImpl> composeServiceEntry : composeEnvironment.getServices()
+                                                                                           .entrySet()) {
+            ComposeServiceImpl service = composeServiceEntry.getValue();
+
+            CheServiceBuildContextImpl buildContext = null;
+            if (service.getBuild() != null) {
```

Comment:
What if you move this section below the declaration of `cheService` then you will be able to avoid buildConext variable e.g.

```java
if (service.getBuild() != null) {
    cheService.setBuild(new CheServiceBuildContextImpl().withContext(service.getBuild().getContext())
                                                        .withDockerfilePath(service.getBuild().getDockerfile())
                                                        .withArgs(service.getBuild().getArgs()))
}
```

Moreover i think that creation of `CheServiceBuildContextImpl` should be replaced with constructor call